lib/cublas/linalg.jl

```diff
@@ -24,6 +24,49 @@ function LinearAlgebra.dot(x::StridedCuArray{T}, y::StridedCuArray{T}) where T<:
     dotc(n, x, y)
 end
 
+function LinearAlgebra.dot(x::StridedCuArray{T1}, y::StridedCuArray{T2}) where {T1,T2}
+    n = length(x)
+    n==length(y) || throw(DimensionMismatch("dot product arguments have lengths $(length(x)) and $(length(y))"))
+
+    T = promote_type(T1, T2)
+    res = CuArray(T[zero(T)])
+    MAX_THREADS = 256
+
+    function kernel(x, y, res, T)
+        index = threadIdx().x
+        thread_stride = blockDim().x
+        block_stride = (length(x)-1) ÷ gridDim().x + 1
+        start = (blockIdx().x - 1) * block_stride + 1
+        stop = blockIdx().x * block_stride
+
+        cache = @cuStaticSharedMem(T, (256#= MAX_THREADS =#,))
+
+        for i in start-1+index:thread_stride:stop
+            @inbounds cache[index] += x[i] * y[i]
+        end
+        sync_threads()
+
+        mid = thread_stride
+        while true
+            mid = (mid - 1) ÷ 2 + 1
+            if index <= mid
+                @inbounds cache[index] += cache[index+mid]
+            end
+            sync_threads()
+            mid == 1 && break
+        end
+
+        if index == 1
+            CUDA.atomic_add!(pointer(res), cache[1])
+        end
+        return
+    end
+    k = @cuda launch=false kernel(x, y, res,T)
+    config = launch_configuration(k.fun)
+    k(x, y, res, T; threads=min(length(x), config.threads, MAX_THREADS), blocks=config.blocks)
```

Comment:
Doesn't `blocks` also need to be clamped?

Reply:
Because I just do thread-level reduction here, so I think `blocks` doesn't need to be clamped?